src/ast/nodes/shared/FunctionNode.ts

```diff
@@ -1,105 +1,98 @@
-import Node, { ForEachReturnExpressionCallback } from '../../Node';
 import FunctionScope from '../../scopes/FunctionScope';
 import BlockScope from '../../scopes/FunctionScope';
-import VirtualObjectExpression from './VirtualObjectExpression';
 import BlockStatement from '../BlockStatement';
 import Identifier from '../Identifier';
-import Pattern from '../Pattern';
 import CallOptions from '../../CallOptions';
 import ExecutionPathOptions from '../../ExecutionPathOptions';
-import { PredicateFunction } from '../../values';
 import { ObjectPath } from '../../variables/VariableReassignmentTracker';
+import { PatternNode } from './Pattern';
+import { GenericExpressionNode, ForEachReturnExpressionCallback, SomeReturnExpressionCallback } from './Expression';
+import { UNKNOWN_OBJECT_EXPRESSION } from '../ObjectExpression';
 
-export default class FunctionNode extends Node {
-  id: Identifier;
-  body: BlockStatement;
-  scope: BlockScope;
-  params: Pattern[];
+export default class FunctionNode extends GenericExpressionNode {
+	id: Identifier;
+	body: BlockStatement;
+	scope: BlockScope;
+	params: PatternNode[];
 
-  prototypeObject: VirtualObjectExpression;
+	bindNode () {
+		this.body.bindImplicitReturnExpressionToScope();
+	}
 
-  bindNode () {
-    this.body.bindImplicitReturnExpressionToScope();
-  }
+	forEachReturnExpressionWhenCalledAtPath (
+		path: ObjectPath,
+		callOptions: CallOptions,
+		callback: ForEachReturnExpressionCallback,
+		options: ExecutionPathOptions
+	) {
+		path.length === 0 &&
+		this.scope.forEachReturnExpressionWhenCalled(
+			callOptions,
+			callback,
+			options
+		);
+	}
 
-  forEachReturnExpressionWhenCalledAtPath (
-    path: ObjectPath,
-    callOptions: CallOptions,
-    callback: ForEachReturnExpressionCallback,
-    options: ExecutionPathOptions
-  ) {
-    path.length === 0 &&
-      this.scope.forEachReturnExpressionWhenCalled(
-        callOptions,
-        callback,
-        options
-      );
-  }
+	hasEffects (options: ExecutionPathOptions) {
+		return this.id && this.id.hasEffects(options);
+	}
 
-  hasEffects (options: ExecutionPathOptions) {
-    return this.id && this.id.hasEffects(options);
-  }
+	hasEffectsWhenAccessedAtPath (path: ObjectPath, options: ExecutionPathOptions) {
+		if (path.length <= 1) {
+			return false;
+		}
+		if (path[0] === 'prototype') {
+			return UNKNOWN_OBJECT_EXPRESSION.hasEffectsWhenAccessedAtPath(path.slice(1), options);
```

Comment:
Could this not just be path.length > 2?

Reply:
I guess you are right. Reusing `UNKNOWN_OBJECT_EXPRESSION` is probably a little over the top here.